contracts/DepegDistribution.sol

```diff
@@ -0,0 +1,242 @@
+
+        // create policy
+        // this will transfer premium amount from this contract to depeg (and keep the commission in this contract)
+        processId = _depegProduct.applyForPolicyWithBundle(
+            protectedWallet,
+            protectedBalance,
+            duration,
+            bundleId);
+
+        emit LogDepegPolicySold(msg.sender, processId, premiumTotalAmount, protectedWallet, protectedBalance);
+    }
+
+    function _collectTokenAndUpdateCommission(
+        address buyer,
+        uint256 protectedBalance,
+        uint256 duration,
+        uint256 bundleId
+    )
+        internal
+        returns (
+            uint256 premiumTotalAmount,
+            uint256 premiumNetAmount,
+            uint256 commissionAmount
+        )
+    {
+        address distributor = msg.sender;
+
+        // calculate amounts
+        (
+            premiumTotalAmount,
+            commissionAmount
+        ) = calculatePrice(distributor, protectedBalance, duration, bundleId);
+
+        premiumNetAmount = premiumTotalAmount - commissionAmount;
+
+        // update distributor book keeping record
+        DistributorInfo storage info = _distributor[distributor];
+        info.commissionBalance += commissionAmount;
+        info.policiesSold += 1;
+        info.updatedAt = block.timestamp;
+
+        // collect total premium amount
+        _token.transferFrom(buyer, address(this), premiumTotalAmount);
+
+        emit LogDistributionInfoUpdated(distributor, commissionAmount, info.commissionBalance, info.policiesSold);
+    }
+
+
+    function calculatePrice(
+        address distributor, 
+        uint256 protectedBalance, 
+        uint256 duration, 
+        uint256 bundleId
+    )
+        public
+        view
+        returns (
+            uint256 premiumTotalAmount,
+            uint256 commissionAmount
+        )
+    {
+        // TODO fetch policy price
+        uint256 sumInsured = _depegRiskpool.calculateSumInsured(protectedBalance);
+        uint256 netPremium = _depegProduct.calculateNetPremium(
+            sumInsured,
+            duration,
+            bundleId);
+
+        uint256 depegPremium = _depegProduct.calculatePremium(netPremium);
+
+        // TODO calculate commission and total premium
+        commissionAmount = calculateCommission(distributor, depegPremium);
+        premiumTotalAmount = depegPremium + commissionAmount;
+    } 
+
+    function calculateCommission(address distributor, uint256 netPremiumAmount)
+        public
+        view
+        returns(uint256 commissionAmount)
+    {
+        uint256 rate = _distributor[distributor].commissionRate;
+        if(rate == 0) {
+            return 0;
+        }
+        
+        return (netPremiumAmount * rate) / (10**DECIMALS - rate);
+    }
+
+    function withdraw(uint256 amount)
+        external
+        onlyOwner()
+    {
+        require(_token.balanceOf(address(this)) >= amount, "ERROR:DST-010:COMMISION_BALANCE_TOO_LARGE");
```

Comment:
Shouldn't this just be `BALANCE_INSUFFICIENT`? Why commission here?

Reply:
what if owner withdraws more tokens than total required to pay pending commision?

Reply:
updated acccordingly